src/build.rs

```diff
@@ -422,66 +466,149 @@ impl BuildQueue {
         // However, if Cargo doesn't run a separate thread, then we'll just wait
         // forever. Therefore, we spawn an extra thread here to be safe.
         let handle = thread::spawn(move || {
-            let mut flags = "-Zunstable-options -Zsave-analysis --error-format=json \
-                             -Zcontinue-parse-after-error".to_owned();
+            #[derive(Debug)]
+            struct CargoOptions {
+                package: Vec<String>,
+                target: Option<String>,
+                lib: bool,
+                bin: Vec<String>,
+                bins: bool,
+                all: bool,
+                exclude: Vec<String>,
+            }
+
+            impl CargoOptions {
+                fn default() -> CargoOptions {
+                    CargoOptions {
+                        package: vec![],
+                        target: None,
+                        lib: false,
+                        bin: vec![],
+                        bins: false,
+                        all: false,
+                        exclude: vec![],
+                    }
+                }
+
+                fn new(config: &Config) -> CargoOptions {
```

Comment:
Initializing those vars inline was quite awkward, since var declaration had to be in a right order to satisfy lifetimes/drop order and was initially not fully defined [here](https://github.com/rust-lang-nursery/rls/pull/409/files/64b2ec3eaed23a7048f14126b59a09e0eae82cb9#diff-9300feef84bdf7e5ab58b05fa4eb0794L438), also the logic in itself was [non-trivial](https://github.com/rust-lang-nursery/rls/pull/409/files/64b2ec3eaed23a7048f14126b59a09e0eae82cb9#diff-9300feef84bdf7e5ab58b05fa4eb0794L453), so with addition of a new mode and conditional initialization of the those config variables, I just went with `new` while filling the blanks with `default` [values](https://github.com/rust-lang-nursery/rls/pull/409/files/64b2ec3eaed23a7048f14126b59a09e0eae82cb9#diff-9300feef84bdf7e5ab58b05fa4eb0794R481), structuring it a little bit more and trying to convey the intent more